File: test/octoladder/sync_test.rb
require "test_helper"
require "octoladder/sync"
require "octoladder/state"
require "octoladder/teams_config"
require "octoladder/github_client"
require "octoladder/config"

class SyncTest < ActiveSupport::TestCase
  def setup
    @config = OctoladderConfig.new("time_zone" => "Asia/Tokyo")
    @teams_config = TeamsConfig.new([
      { "org" => "acme",   "team_slug" => "platform" },
      { "org" => "acme",   "team_slug" => "infra" }
    ])
    @client = GithubClient.new(token: "test-token")
    @now = Time.utc(2026, 4, 27, 17, 0, 0)
  end

  def stub_team_members(org, slug, members)
    stub_request(:get, "https://api.github.com/orgs/#{org}/teams/#{slug}/members?per_page=100")
      .to_return(
        status: 200,
        headers: { "Content-Type" => "application/json" },
        body: members.to_json
      )
  end

  def stub_search(login, items: [], total_count: nil)
    total_count ||= items.size
    stub_request(:get, %r{api\.github\.com/search/issues})
      .with(query: hash_including("q" => /author:#{login}/))
      .to_return(
        status: 200,
        headers: { "Content-Type" => "application/json" },
        body: { total_count: total_count, items: items }.to_json
      )
  end

  def pr_item(id, repo:, merged_at:)
    {
      id: id,
      html_url: "https://github.com/#{repo}/pull/#{id}",
      repository_url: "https://api.github.com/repos/#{repo}",
      pull_request: { merged_at: merged_at }
    }
  end

  def run_sync(state)
    Sync.new(
      state: state,
      teams_config: @teams_config,
      github_client: @client,
      config: @config,
      now: @now
    ).call
  end

  # ---- empty state backfill ----

  test "empty state backfill adds users and PRs from the anchor" do
    stub_team_members("acme", "platform", [
      { id: 1, login: "alice", avatar_url: "https://example.com/a.png" }
    ])
    stub_team_members("acme", "infra", [])

    expected_from = "2025-01-01T00:00:00Z"
    expected_to = "2026-04-27T16:59:59Z"
    expected_q = "is:pr is:merged is:public author:alice merged:#{expected_from}..#{expected_to}"
    stub_request(:get, "https://api.github.com/search/issues")
      .with(query: hash_including("q" => expected_q))
      .to_return(
        status: 200,
        headers: { "Content-Type" => "application/json" },
        body: { total_count: 1, items: [
          pr_item(100, repo: "acme/widget", merged_at: "2026-04-20T09:00:00Z")
        ] }.to_json
      )

    state = State.new
    run_sync(state)

    assert_equal 1, state.users.size
    assert_equal "alice", state.users.first["login"]
    assert_equal true, state.users.first["active"]
    assert_equal [ "acme/platform" ], state.users.first["team_keys"]

    assert_equal 1, state.pull_requests.size
    pr = state.pull_requests.first
    assert_equal 100, pr["github_id"]
    assert_equal "alice", pr["author_login"]
    assert_equal "2026-04-20T09:00:00Z", pr["merged_at"]
    assert_equal "acme/widget", pr["repo_full_name"]

    assert_equal @now, state.synced_at
    assert_equal Date.new(2025, 1, 1), state.backfill_anchor
  end

  # ---- incremental fetch ----

  test "incremental fetch starts one day before the latest recorded PR" do
    stub_team_members("acme", "platform", [
      { id: 1, login: "alice", avatar_url: "https://example.com/a.png" }
    ])
    stub_team_members("acme", "infra", [])

    expected_from = "2026-04-19T09:00:00Z"
    expected_to = "2026-04-27T16:59:59Z"
    stub_request(:get, "https://api.github.com/search/issues")
      .with(query: hash_including("q" => "is:pr is:merged is:public author:alice merged:#{expected_from}..#{expected_to}"))
      .to_return(
        status: 200,
        headers: { "Content-Type" => "application/json" },
        body: { total_count: 0, items: [] }.to_json
      )

    state = State.new(
      synced_at: Time.utc(2026, 4, 20),
      backfill_anchor: Date.new(2025, 1, 1),
      users: [ {
        "github_id" => 1, "login" => "alice", "avatar_url" => "x",
        "team_keys" => [ "acme/platform" ], "active" => true
      } ],
      pull_requests: [ {
        "github_id" => 100, "author_login" => "alice",
        "merged_at" => "2026-04-20T09:00:00Z",
        "html_url" => "x", "repo_full_name" => "acme/widget"
      } ]
    )
    run_sync(state)

    assert_requested :get, %r{api\.github\.com/search/issues},
      query: hash_including("q" => /merged:#{expected_from}/)
  end

  # ---- user join ----

  test "newly added team member is recorded as active" do
    stub_team_members("acme", "platform", [
      { id: 1, login: "alice", avatar_url: "x" },
      { id: 2, login: "bob", avatar_url: "y" }
    ])
    stub_team_members("acme", "infra", [])
    stub_search("alice")
    stub_search("bob")

    state = State.new(
      synced_at: Time.utc(2026, 4, 20),
      backfill_anchor: Date.new(2025, 1, 1),
      users: [ {
        "github_id" => 1, "login" => "alice", "avatar_url" => "x",
        "team_keys" => [ "acme/platform" ], "active" => true
      } ]
    )
    run_sync(state)

    bob = state.users.find { |u| u["login"] == "bob" }
    refute_nil bob
    assert_equal true, bob["active"]
    assert_equal [ "acme/platform" ], bob["team_keys"]
  end

  # ---- user leave ----

  test "user who left every tracked team is deactivated, PRs preserved" do
    stub_team_members("acme", "platform", []) # alice gone
    stub_team_members("acme", "infra", [])

    state = State.new(
      synced_at: Time.utc(2026, 4, 20),
      backfill_anchor: Date.new(2025, 1, 1),
      users: [ {
        "github_id" => 1, "login" => "alice", "avatar_url" => "x",
        "team_keys" => [ "acme/platform" ], "active" => true
      } ],
      pull_requests: [ {
        "github_id" => 100, "author_login" => "alice",
        "merged_at" => "2026-04-20T09:00:00Z",
        "html_url" => "x", "repo_full_name" => "acme/widget"
      } ]
    )
    run_sync(state)

    alice = state.users.find { |u| u["login"] == "alice" }
    assert_equal false, alice["active"]
    assert_empty alice["team_keys"]
    assert_equal 1, state.pull_requests.size # unchanged
    assert_not_requested :get, %r{api\.github\.com/search/issues}
  end

  # ---- team removal ----

  test "removing a team from config stops fetching for users only in that team" do
    teams_config = TeamsConfig.new([ { "org" => "acme", "team_slug" => "platform" } ]) # infra removed
    stub_team_members("acme", "platform", [])

    state = State.new(
      synced_at: Time.utc(2026, 4, 20),
      backfill_anchor: Date.new(2025, 1, 1),
      users: [ {
        "github_id" => 1, "login" => "alice", "avatar_url" => "x",
        "team_keys" => [ "acme/infra" ], "active" => true
      } ],
      pull_requests: []
    )

    Sync.new(
      state: state, teams_config: teams_config,
      github_client: @client, config: @config, now: @now
    ).call

    assert_equal false, state.users.first["active"]
    assert_not_requested :get, %r{api\.github\.com/search/issues}
    assert_equal [ { "org" => "acme", "slug" => "platform" } ], state.teams
  end

  # ---- multi-team membership ----

  test "user belonging to multiple tracked teams gets sorted team_keys" do
    stub_team_members("acme", "platform", [
      { id: 1, login: "alice", avatar_url: "x" }
    ])
    stub_team_members("acme", "infra", [
      { id: 1, login: "alice", avatar_url: "x" }
    ])
    stub_search("alice")

    state = State.new
    run_sync(state)

    assert_equal [ "acme/infra", "acme/platform" ], state.users.first["team_keys"]
  end

  # ---- PR dedup ----

  test "PRs returned again on subsequent sync are deduped by github_id" do
    stub_team_members("acme", "platform", [
      { id: 1, login: "alice", avatar_url: "x" }
    ])
    stub_team_members("acme", "infra", [])

    stub_request(:get, %r{api\.github\.com/search/issues})
      .to_return(
        status: 200,
        headers: { "Content-Type" => "application/json" },
        body: { total_count: 1, items: [
          pr_item(100, repo: "acme/widget", merged_at: "2026-04-20T09:00:00Z")
        ] }.to_json
      )

    state = State.new(
      synced_at: Time.utc(2026, 4, 20),
      backfill_anchor: Date.new(2025, 1, 1),
      users: [ {
        "github_id" => 1, "login" => "alice", "avatar_url" => "x",
        "team_keys" => [ "acme/platform" ], "active" => true
      } ],
      pull_requests: [ {
        "github_id" => 100, "author_login" => "alice",
        "merged_at" => "2026-04-20T09:00:00Z",
        "html_url" => "x", "repo_full_name" => "acme/widget"
      } ]
    )
    run_sync(state)

    assert_equal 1, state.pull_requests.size
  end
end

File: test/sync.test.ts
import nock from "nock";
import { afterAll, afterEach, beforeAll, describe, expect, it, vi } from "vitest";
import { OctoladderConfig } from "../src/config.js";
import { State } from "../src/state.js";
import { Sync, type Logger } from "../src/sync.js";
import { TeamsConfig } from "../src/teams-config.js";
import { makeClient, makePullRequest, makeUser } from "./helpers.js";

beforeAll(() => nock.disableNetConnect());
afterEach(() => nock.cleanAll());
afterAll(() => nock.enableNetConnect());

const NOW = new Date("2026-04-27T17:00:00Z");

function makeSync(
  state: State,
  teamsConfig?: TeamsConfig,
  config?: OctoladderConfig,
  logger?: Logger,
): Sync {
  return new Sync({
    state,
    teamsConfig: teamsConfig ?? defaultTeamsConfig(),
    githubClient: makeClient(),
    config: config ?? new OctoladderConfig({ time_zone: "Asia/Tokyo" }),
    now: NOW,
    logger,
  });
}

function defaultTeamsConfig(): TeamsConfig {
  return new TeamsConfig([
    { org: "acme", team_slug: "platform" },
    { org: "acme", team_slug: "infra" },
  ]);
}

function stubMembers(org: string, slug: string, members: { id: number; login: string; avatar_url: string }[]): void {
  nock("https://api.github.com")
    .get(`/orgs/${org}/teams/${slug}/members`)
    .query({ per_page: 100 })
    .reply(200, members);
}

function stubSearch(login: string, items: SearchItem[] = [], totalCount?: number): void {
  nock("https://api.github.com")
    .get("/search/issues")
    .query((q) => typeof q["q"] === "string" && q["q"].includes(`author:${login}`))
    .reply(200, { total_count: totalCount ?? items.length, items });
}

interface SearchItem {
  id: number;
  html_url: string;
  repository_url: string;
  pull_request: { merged_at: string };
}

function prItem(id: number, repo: string, mergedAt: string): SearchItem {
  return {
    id,
    html_url: `https://github.com/${repo}/pull/${id}`,
    repository_url: `https://api.github.com/repos/${repo}`,
    pull_request: { merged_at: mergedAt },
  };
}

describe("Sync", () => {
  it("empty state backfill adds users and PRs from the anchor", async () => {
    stubMembers("acme", "platform", [
      { id: 1, login: "alice", avatar_url: "https://example.com/a.png" },
    ]);
    stubMembers("acme", "infra", []);
    stubSearch("alice", [prItem(100, "acme/widget", "2026-04-20T09:00:00Z")]);

    const state = new State();
    await makeSync(state).call();

    expect(state.users).toHaveLength(1);
    const alice = state.users[0]!;
    expect(alice.login).toBe("alice");
    expect(alice.active).toBe(true);
    expect(alice.team_keys).toEqual(["acme/platform"]);

    expect(state.pullRequests).toHaveLength(1);
    const pr = state.pullRequests[0]!;
    expect(pr).toMatchObject({
      github_id: 100,
      author_login: "alice",
      merged_at: "2026-04-20T09:00:00Z",
      repo_full_name: "acme/widget",
    });

    expect(state.syncedAt).toEqual(NOW);
    expect(state.backfillAnchor).toEqual(new Date("2025-01-01T00:00:00Z"));
  });

  it("incremental fetch starts one day before the latest recorded PR", async () => {
    stubMembers("acme", "platform", [{ id: 1, login: "alice", avatar_url: "x" }]);
    stubMembers("acme", "infra", []);

    const expectedFromIso = "2026-04-19T09:00:00Z";
    nock("https://api.github.com")
      .get("/search/issues")
      .query((q) => typeof q["q"] === "string" && q["q"].includes(`merged:${expectedFromIso}`))
      .reply(200, { total_count: 0, items: [] });

    const state = new State({
      syncedAt: new Date("2026-04-20T00:00:00Z"),
      backfillAnchor: new Date("2025-01-01T00:00:00Z"),
      users: [makeUser()],
      pullRequests: [makePullRequest()],
    });
    await makeSync(state).call();

    expect(nock.isDone()).toBe(true);
  });

  it("newly added team member is recorded as active", async () => {
    stubMembers("acme", "platform", [
      { id: 1, login: "alice", avatar_url: "x" },
      { id: 2, login: "bob", avatar_url: "y" },
    ]);
    stubMembers("acme", "infra", []);
    stubSearch("alice");
    stubSearch("bob");

    const state = new State({
      syncedAt: new Date("2026-04-20T00:00:00Z"),
      backfillAnchor: new Date("2025-01-01T00:00:00Z"),
      users: [makeUser()],
    });
    await makeSync(state).call();

    const bob = state.users.find((u) => u.login === "bob");
    expect(bob).toBeDefined();
    expect(bob!.active).toBe(true);
    expect(bob!.team_keys).toEqual(["acme/platform"]);
  });

  it("user who left every tracked team is deactivated, PRs preserved", async () => {
    stubMembers("acme", "platform", []);
    stubMembers("acme", "infra", []);

    const state = new State({
      syncedAt: new Date("2026-04-20T00:00:00Z"),
      backfillAnchor: new Date("2025-01-01T00:00:00Z"),
      users: [makeUser()],
      pullRequests: [makePullRequest()],
    });
    await makeSync(state).call();

    const alice = state.users.find((u) => u.login === "alice")!;
    expect(alice.active).toBe(false);
    expect(alice.team_keys).toEqual([]);
    expect(state.pullRequests).toHaveLength(1);
  });

  it("removing a team from config stops fetching for users only in that team", async () => {
    stubMembers("acme", "platform", []);

    const teamsConfig = new TeamsConfig([{ org: "acme", team_slug: "platform" }]);
    const state = new State({
      syncedAt: new Date("2026-04-20T00:00:00Z"),
      backfillAnchor: new Date("2025-01-01T00:00:00Z"),
      users: [makeUser({ team_keys: ["acme/infra"] })],
    });
    await makeSync(state, teamsConfig).call();

    expect(state.users[0]!.active).toBe(false);
    expect(state.teams).toEqual([{ org: "acme", slug: "platform" }]);
  });

  it("user belonging to multiple tracked teams gets sorted team_keys", async () => {
    stubMembers("acme", "platform", [{ id: 1, login: "alice", avatar_url: "x" }]);
    stubMembers("acme", "infra", [{ id: 1, login: "alice", avatar_url: "x" }]);
    stubSearch("alice");

    const state = new State();
    await makeSync(state).call();

    expect(state.users[0]!.team_keys).toEqual(["acme/infra", "acme/platform"]);
  });

  it("forwards configured min_stars to the search query", async () => {
    stubMembers("acme", "platform", [{ id: 1, login: "alice", avatar_url: "x" }]);
    stubMembers("acme", "infra", []);

    nock("https://api.github.com")
      .get("/search/issues")
      .query((q) => typeof q["q"] === "string" && q["q"].includes("stars:>=50"))
      .reply(200, { total_count: 0, items: [] });

    const state = new State();
    const config = new OctoladderConfig({ time_zone: "Asia/Tokyo", min_stars: 50 });
    await makeSync(state, undefined, config).call();

    expect(nock.isDone()).toBe(true);
  });

  it("user exceeding the search cap is skipped with a warning, others still processed", async () => {
    stubMembers("acme", "platform", [
      { id: 1, login: "alice", avatar_url: "x" },
      { id: 2, login: "bob", avatar_url: "y" },
    ]);
    stubMembers("acme", "infra", []);
    stubSearch("alice", [], 1234);
    stubSearch("bob", [prItem(200, "acme/widget", "2026-04-20T09:00:00Z")]);

    const warn = vi.fn();
    const state = new State();
    await makeSync(state, undefined, undefined, { warn }).call();

    expect(state.pullRequests).toHaveLength(1);
    expect(state.pullRequests[0]!.author_login).toBe("bob");
    expect(state.syncedAt).toEqual(NOW);
    expect(state.backfillAnchor).toEqual(new Date("2025-01-01T00:00:00Z"));
    expect(warn).toHaveBeenCalledTimes(1);
    expect(warn.mock.calls[0]![0]).toMatch(/^Skipping alice: /);
    expect(warn.mock.calls[0]![0]).toContain("1234");
  });

  it("non-truncation errors propagate and abort the sync", async () => {
    stubMembers("acme", "platform", [{ id: 1, login: "alice", avatar_url: "x" }]);
    stubMembers("acme", "infra", []);
    nock("https://api.github.com")
      .get("/search/issues")
      .query(true)
      .reply(500, { message: "boom" });

    const warn = vi.fn();
    const state = new State();
    await expect(makeSync(state, undefined, undefined, { warn }).call()).rejects.toThrow();
    expect(warn).not.toHaveBeenCalled();
  });

  it("PRs returned again on subsequent sync are deduped by github_id", async () => {
    stubMembers("acme", "platform", [{ id: 1, login: "alice", avatar_url: "x" }]);
    stubMembers("acme", "infra", []);

    nock("https://api.github.com")
      .get("/search/issues")
      .query(true)
      .reply(200, {
        total_count: 1,
        items: [prItem(100, "acme/widget", "2026-04-20T09:00:00Z")],
      });

    const state = new State({
      syncedAt: new Date("2026-04-20T00:00:00Z"),
      backfillAnchor: new Date("2025-01-01T00:00:00Z"),
      users: [makeUser()],
      pullRequests: [makePullRequest()],
    });
    await makeSync(state).call();

    expect(state.pullRequests).toHaveLength(1);
  });
});
